packages/@lwc/engine-core/src/framework/hooks.ts

```diff
@@ -37,6 +38,29 @@ function setElementShadowToken(elm: Element, token: string | undefined) {
     (elm as any).$shadowToken$ = token;
 }
 
+function assertNotSyntheticComposedWithinNative(vm: VM) {
+    const isSynthetic =
+        vm.renderMode === RenderMode.Shadow && vm.shadowMode === ShadowMode.Synthetic;
+    if (!isSynthetic) {
+        return;
+    }
+    let ancestor = vm.owner;
+    while (!isNull(ancestor)) {
+        const ancestorIsNative =
+            ancestor.renderMode === RenderMode.Shadow && ancestor.shadowMode === ShadowMode.Native;
+        // Perf: wrap in an `if` to avoid creating the string unnecessarily
+        if (ancestorIsNative) {
```

Comment:
Let's remove the perf micro-optimization and wrap the `assertNotSyntheticComposedWithinNative` in the `process.env.NODE_ENV` check.